File: P3/java/deepspace/src/deepspace/Weapon.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

class Weapon{
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String n, final WeaponType w, int u) { //NO SE SI SE PONE EL FINAL
        name = n;
        type = w;
        uses = u;
    }
    
    Weapon(final Weapon other) {
        name = other.name;
        type = other.type;
        uses = other.uses;
    }
    
    WeaponToUI getUIversion(){
        return new WeaponToUI(this);
    }
    
    public String toString(){
        return getUIversion().toString();
    }
    
    WeaponType getType() {
        return type;
    }
    
    int getUses() {
        return uses;
    }
    
    public float power() {
        return type.getPower();
    }
    
    public float useIt() {
        if(uses > 0){
            uses -= 1;
            return power();
        }else
            return 1;
    }
}

File: P2/ruby/Weapon.rb
# encoding: UTF-8

require_relative 'WeaponType'
require_relative 'WeaponToUI'

module Deepspace
	class Weapon
		def initialize (n, t, u) # Weapon(n : string, t : WeaponType, u : int)
			@name = n
			@type = t
			@uses = u
		end
		
		attr_reader :name
		attr_reader :type
		attr_reader :uses
		
		def self.newCopy (other)
			new(other.name, other.type, other.uses)
		end

		def getUIVersion
			WeaponToUI.new(self)
		end
		
		def to_s
			"name: #{@name}, type: [#{@type}], uses: #{@uses}"
		end

		def power
			@type.power
		end
		
		def useIt
			if @uses > 0
				@uses -= 1
				power
			else
				1.0
			end
		end
	end
end

if $0 == __FILE__
	puts Deepspace::Weapon.new(1,2,3).to_s
	puts Deepspace::Weapon.new(1, Deepspace::WeaponType::LASER, 3).getUIVersion.to_s
end